src/lib.rs

```diff
@@ -563,13 +564,27 @@ impl FileDialog {
                 }
 
                 if response.double_clicked() {
-                  command = Some(match self.dialog_type == DialogType::SaveFile {
-                    true => match info.dir {
-                      true => Command::OpenSelected,
-                      false => Command::Save(info.clone()),
-                    },
-                    false => Command::Open(info.clone()),
-                  });
+                  match self.dialog_type {
+                    DialogType::SelectFolder => {
+                      // always open folder on double click, otherwise SelectFolder cant enter subfolders
+                      command = Some(Command::OpenSelected);
+                    }
+                    // open or save file only if name matches filter
+                    DialogType::OpenFile => {
+                      if info.dir {
+                        command = Some(Command::OpenSelected);
+                      } else if (self.filename_filter)(self.filename_edit.as_str()) {
+                        command = Some(Command::Open(info.clone()));
+                      }
+                    }
+                    DialogType::SaveFile => {
+                      if info.dir {
+                        command = Some(Command::OpenSelected);
+                      } else if (self.filename_filter)(self.filename_edit.as_str()) {
+                        command = Some(Command::Save(info.clone()));
+                      }
+                    }
```

Comment:
`OpenFile` and `SaveFile` look exactly the same. They could be combined.

```Rust
                    DialogType::OpenFile | DialogType::SaveFile => {
                      // open or save file only if name matches filter
                      if info.dir {
                        command = Some(Command::OpenSelected);
                      } else if (self.filename_filter)(self.filename_edit.as_str()) {
                        command = Some(Command::Open(info.clone()));
                      }
                    }
```

Reply:
No, OpenFile does Command::Open and SaveFile does Command::Save